### PR TITLE
feat(Page): add Page type

### PR DESCRIPTION
**Pull request by traversng**
add Page tests
separate validations from Shapes

close #qc4j7

### CODE REVIEW

File: src/utils/__tests__/validations.spec.ts [hunk 1]
```diff
@@ -222,5 +223,34 @@ describe('validate', () => {
...
+        route: 'someplace'
+      })
+
+      const errors = validateShape(home, [Author123])
+      expect(errors).toHaveLength(0)
+    })
+
+    test('A Page should not be able to embed another page', () => {
+      const home = new Page({
+        fields: { someField: { type: 'page' } },
```
Comment (Hendrixer) [inline src/utils/__tests__/validations.spec.ts]: for this test, you want the type to be `types.shape` and a `ref` to point to the API ID of a page

File: tslint.json [hunk 2]
```diff
@@ -10,6 +10,7 @@
     "ordered-imports": false,
     "only-arrow-functions": false,
     "object-literal-sort-keys": false,
-    "no-console": false
+    "no-console": false,
+    "no-unused-expression": false
```
Comment (Hendrixer) [inline tslint.json]: hmmmmmmmm

Comment (traversng) [reply tslint.json]: Yea, this is nasty. Ill refactor the test to something that'll pass this rule.

File: src/utils/__tests__/validations.spec.ts (hunk 1 above)
Comment (traversng) [reply src/utils/__tests__/validations.spec.ts]: Ok

File: src/page/page.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    if (this.isSingleRoute) return routeParamsStore
+    pathToRegexp(route, routeParamsStore)
+    return routeParamsStore
+  }
+
+  public routeIsSingle(route: string): boolean {
+    return !route.includes('/:')
+  }
+
+  public normalizeFields(fields: IFields): IFields {
```
[the remark this answers is not in the document]
Comment (traversng) [reply src/page/page.ts]: Ok, Ill make an edit

File: src/page/page.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  constructor(options: IPageOptions) {
+    if (!isString(options.apiId)) throw new Error('Page API ID must be string')
+
+    if (!isString(options.route)) throw new Error('Page route must be string')
+
+    this.apiId = options.apiId
+    this.name = options.name || options.apiId
+    this.fields = this.normalizeFields(options.fields)
+    this.routeParams = []
```
[the remark this answers is not in the document]
Comment (traversng) [reply src/page/page.ts]: can you explain? right now routeParams is just a store for the eventual parameters (either empty array if single route or the result of regexp)

[the remark this answers is not in the document]
Comment (Hendrixer) [reply src/page/page.ts]: those params come from `options.route` after its validated. Using the same regexp function, you need to store the params in this array

File: src/page/page.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      })
+      .map((obj: any) => obj.name)
+  }
+
+  public exposeRouteParams(route: string, routeParamsStore: any[]): any[] {
+    if (this.isSingleRoute) return routeParamsStore
+    pathToRegexp(route, routeParamsStore)
+    return routeParamsStore.map(param => {
+      const { name } = param
+      return { name }
```
Comment (Hendrixer) [inline src/page/page.ts]: all we need is an array of strings, not an array of objects

**Pull request merged**